### ISSUE TITLE
Can't filter STIX2 content by type

### ISSUE THREAD

**Issue by leveled** [mitre/cti#36]
Not sure if this is an a bug in the CTI database or with the stix2 library, but I am unable to filter on type. For whatever reason every other filter I tested was working and type is not. 

```
>>> f1 = Filter("name", "=", ".bash_profile and .bashrc")
>>> techniques = fs.query(f1)
>>> print(techniques)
[AttackPattern(type=u'attack-pattern', id=u'attack-pattern--01df3350-ce05-4bdf-bdf8-0a919a66d4a8', created_by_ref='identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5', created='2017-12-14T16:46:06.044Z', modified='2018-10-17T00:14:20.652Z', name=u'.bash_profile and .bashrc', description=u"<code>~/.bash_profile</code> and <code>~/.bashrc</code> are executed in a user's context when a new shell opens or when a user logs in so that their environment is set correctly. <code>~/.bash_profile</code> is executed for login shells and <code>~/.bashrc</code> is executed for interactive non-login shells. This means that when a user logs in (via username and password) to the console (either locally or remotely via something like SSH), <code>~/.bash_profile</code> is executed before the initial command prompt is returned to the user. After that, every time a new shell is opened, <code>~/.bashrc</code> is executed. This allows users more fine grained control over when they want certain commands executed.\n\nMac's Terminal.app is a little different in that it runs a login shell by default each time a new terminal window is opened, thus calling <code>~/.bash_profile</code> each time instead of <code>~/.bashrc</code>.\n\nThese files are meant to be written to by the local user to configure their own environment; however, adversaries can also insert code into these files to gain persistence each time a user logs in or opens a new shell  (Citation: amnesia malware).", kill_chain_phases=[KillChainPhase(kill_chain_name=u'mitre-attack', phase_name=u'persistence')], external_references=[ExternalReference(source_name=u'mitre-attack', url=u'https://attack.mitre.org/techniques/T1156', external_id=u'T1156'), ExternalReference(source_name=u'amnesia malware', description=u'Claud Xiao, Cong Zheng, Yanhui Jia. (2017, April 6). New IoT/Linux Malware Targets DVRs, Forms Botnet. Retrieved February 19, 2018.', url=u'https://researchcenter.paloaltonetworks.com/2017/04/unit42-new-iotlinux-malware-targets-dvrs-forms-botnet/')], object_marking_refs=['marking-definition--fa42a846-8d90-4e51-bc29-71d5b4802168'], x_mitre_data_sources=[u'File monitoring', u'Process Monitoring', u'Process command-line parameters', u'Process use of network'], x_mitre_detection=u'While users may customize their <code>~/.bashrc</code> and <code>~/.bash_profile</code> files , there are only certain types of commands that typically appear in these files. Monitor for abnormal commands such as execution of unknown programs, opening network sockets, or reaching out across the network when user profiles are loaded during the login process.', x_mitre_permissions_required=[u'User', u'Administrator'], x_mitre_platforms=[u'Linux', u'macOS'], x_mitre_version=u'1.0')]

>>> f1 = Filter("type", "=", "attack-pattern")
>>> techniques = fs.query(f1)
>>> print(techniques)
[]
```

I didn't alter the CTI data from the repo here, any idea how I can debug or fix this issue?

**Comment by jburns12** [comment]
Hi @leveled - hopefully I can help you out with this. I'm having a hard time recreating the issue, but I can think of a few things that might be going on. If you don't mind answering a couple questions that would help troubleshooting a bit:

1. Do you have the latest version of the ATT&CK content? I ask because originally (right when the October 2018 content was released) I realized there was an improperly formatted attack-pattern in enterprise-attack. I committed a fix to this repo a few hours later, but was thinking that maybe if that attack-pattern was in your content it could cause an issue.

2. If that's not the problem, can you tell me what version of the stix2 library you're using?

Thanks!

**Comment by leveled** [comment]
I created a clean folder with a Python 3 virtual environment and was still getting the error 

```
(venv)➜  attack_update_test ls
ATTACK venv
(venv)➜  attack_update_test python3
Python 3.6.5 (default, Mar 31 2018, 13:09:02)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import stix2
>>> from stix2 import FileSystemStore
>>> fs = FileSystemStore('./ATTACK')
>>> from stix2 import Filter
>>> filt = Filter('type', '=', 'attack-pattern')
>>> techniques = fs.query([filt])
>>> print(techniques)
[]
>>> quit()
(venv)➜  attack_update_test pip3 show stix2
---
Metadata-Version: 2.1
Name: stix2
Version: 1.0.2
Summary: Produce and consume STIX 2 JSON content
Home-page: https://github.com/oasis-open/cti-python-stix2
Author: OASIS Cyber Threat Intelligence Technical Committee
Author-email: cti-users@lists.oasis-open.org
License: BSD
Location: /Users/adi/github/attack_update_test/venv/lib/python3.6/site-packages
Requires: pytz, simplejson, stix2-patterns, six, python-dateutil, requests
```

Just in case it was a strange MacOS bug I checked it on a Kali VM I had running again, and got the same issue

```
(venv) root@kali:~/attack_update# python3
Python 3.6.6 (default, Jun 27 2018, 14:44:17)
[GCC 8.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import stix2
>>> from stix2 import FileSystemStore
>>> fs = FileSystemStore('./ATTACK')
>>> from stix2 import Filter
>>> filt = Filter('type', '=', 'attack-pattern')
>>> techniques = fs.query([filt])
>>> print(techniques)
[]
>>>
```

I got the same error when I tried this in a Python 2 environment. Filters on other fields seem to work just fine. Perhaps you can tell me you managed to _not_ get the error and I can try to replicate that.

**Comment by emmanvg** [comment]
Hi @leveled, it took a while to figure out the problem you were experiencing.

I was looking back at the commit history for `stix2` and the `FileSystemSource` query code has not changed in a while. Now, this data source makes an assumption that you have a parent directory followed by directories that have the type as the name (example below). Similar to what you see in this repo. In the past, only ATTACK (or what is now attack-enterprise) data was hosted here, but now there are other data sources hosted here as well.

The FileSystemSource will recursively go deeper to find data, but another thing related to the funny results you were obtaining was that a "type" Filter is treated differently by the FileSystemSource. It essentially serves as a hint to search in a specific directory. Since below cti/ there is no attack-pattern/ directory it returns an empty list. On the hand, Filter("name", "=", ".bash_profile and .bashrc") will search naively go through all directories.

```
parent/
     attack-pattern/
     campaign/
     ... (so on for each type including custom objects)
```

I believe you encountered this problem because of the USAGE.md (which needs to be updated). Now to overcome this you have various options:

Case 1: I only need to query the enterprise-attack sources

``` python
import stix2

enterprise_attack_ds = stix2.FileSystemStore("./cti/enterprise-attack")

f1 = stix2.Filter("name", "=", ".bash_profile and .bashrc")
f2 = stix2.Filter("type", "=", "attack-pattern")

print(len(enterprise_attack_ds.query(f1)))  # Expect 1
print(len(enterprise_attack_ds.query(f2)))  # Expect 223

```

Case 2: I need to query multiple data sources (e.g., mobile, enterprise, pre)

``` python
import stix2

enterprise_attack_ds = stix2.FileSystemSource("./cti/enterprise-attack")
pre_attack_ds = stix2.FileSystemSource("./cti/pre-attack")
mobile_attack_ds = stix2.FileSystemSource("./cti/mobile-attack")
composite_ds = stix2.CompositeDataSource()
composite_ds.add_data_sources([enterprise_attack_ds, pre_attack_ds, mobile_attack_ds])

f1 = stix2.Filter("name", "=", ".bash_profile and .bashrc")
f2 = stix2.Filter("type", "=", "attack-pattern")

print(len(composite_ds.query(f1)))  # Expect 1
print(len(composite_ds.query(f2)))  # Expect 478

```

Hope this helps!

**Comment by leveled** [comment]
Hey @emmanvg, that cleared up my issue with using a local datastore but I'm also having trouble with the MITRE hosted TAXII server. I'm not sure if I should bring this up in a separate issue but I'll detail the problem below as the solution may be similar. 

I'm unable to use the documentation listed on [this MITRE blog post](https://www.mitre.org/capabilities/cybersecurity/overview/cybersecurity-blog/attck%E2%84%A2-content-available-in-stix%E2%84%A2-20-via) to do any sorts of queries via the TAXII server.

The collections appear to be sourced correctly

```>>> import stix2
>>> from stix2 import TAXIICollectionSource
>>> from taxii2client import Server
>>> server = Server("https://cti-taxii.mitre.org/taxii/")
>>> api_root = server.api_roots[0]
>>> for collection in api_root.collections:
...     print(collection.title + ": "+ collection.id)
...
Enterprise ATT&CK: 95ecc380-afe9-11e4-9b6c-751b66dd541e
PRE-ATT&CK: 062767bd-02d2-4b72-84ba-56caef0f8658
Mobile ATT&CK: `2f669986-b40b-4423-b720-4396ca6a462b```

However I'm unable to perform any queries on a TAXIICollectionSource 

```>>> from taxii2client import Collection
>>> collection = Collection("https://cti-taxii.mitre.org/stix/collections/95ecc380-afe9-11e4-9b6c-751b66dd541e")
>>> tc_source = TAXIICollectionSource(collection)
>>> f1 = stix2.Filter("type","=","attack-pattern")
>>> tc_source.query(f1)
[]
>>> f2 = stix2.Filter("name","=",".bash_profile and .bashrc")
>>> tc_source.query(f2)
[]```

**Comment by jburns12** [comment]
Hi @leveled - I played around with this a bit and realized the issue is adding a slash at the end of the argument to the Collection API.

For example:
```
collection = Collection("https://cti-taxii.mitre.org/stix/collections/95ecc380-afe9-11e4-9b6c-751b66dd541e/")
```
@emmanvg - is this expected behavior?

**Comment by emmanvg** [comment]
@jburns12, this is currently an open issue in the `taxii2-client` repo. https://github.com/oasis-open/cti-taxii-client/issues/50

**Comment by leveled** [comment]
Alright that seems to work, thanks for the help!
